politician_analysis/src/scraper.py

```diff
@@ -0,0 +1,126 @@
+SORT_BY_POPULAR = 0
+SORT_BY_RECENT = 1
+
+def find_value(html, key, num_chars=2, separator='"'):
+    pos_begin = html.find(key) + len(key) + num_chars
+    pos_end = html.find(separator, pos_begin)
+    return html[pos_begin: pos_end]
+
+
+def ajax_request(session, endpoint, version, api_key, retries=5, sleep=20):
+    
+    component = endpoint['commandMetadata']['webCommandMetadata']['apiUrl']
+    
+    url = f'https://www.youtube.com{component}'
+    data = {'context': {'client': {'userAgent': USER_AGENT, 'clientName': 'WEB', 'clientVersion': version},
+                        'clickTracking': {'clickTrackingParams': endpoint['clickTrackingParams']}},
+            "continuation": endpoint['continuationCommand']['token']}
+
+    for _ in range(retries):
+        response = session.post(url, params={'key': api_key}, json=data)
+        if response.status_code == 200:
+            return response.json()
+        elif response.status_code in [403, 413]:
+            return {}
+        else:
+            time.sleep(sleep)
+
+def search_dict(partial, search_key):
+    stack = [partial]
+    while stack:
+        current_item = stack.pop()
+        if isinstance(current_item, dict):
+            for key, value in current_item.items():
+                if key == search_key:
+                    yield value
+                else:
+                    stack.append(value)
+        elif isinstance(current_item, list):
+            for value in current_item:
+                stack.append(value)
+
+
+def download_comments(youtube_id, sort_by=SORT_BY_RECENT, sleep=.1):
+    
+    """Download comments from youtube URL, send ajax request for dynamic webpages.
+    Load the response JSON file and sorted the comments if it's not sorted by popularity.
+    Lastly, save comment information as a dictionary.
+    
+
+    Parameters:
+    youtube_id : youtube id for the selected video.
+    sort_by : Has SORT_BY_RECENT and SORT_BY_POPULAR based on customers settings.
+    sleep : wait for a given number of seconds.
+
+    Returns:
+    comment data: A dictionary with selected comments information.
+    """
+    
+    
+    session = requests.Session()
+    session.headers['User-Agent'] = USER_AGENT
+
+    response = session.get(YOUTUBE_URL.format(youtube_id=youtube_id))
+
+    if 'uxe=' in response.request.url:
+        session.cookies.set('CONSENT', 'YES+cb', domain='.youtube.com')
+        response = session.get(YOUTUBE_URL.format(youtube_id=youtube_id))
+
+    html = response.text
+    version = find_value(html, 'client.version', 4, '\'')
+    api_key = find_value(html, 'INNERTUBE_API_KEY', 3)
+
+    data = json.loads(find_value(html, 'var ytInitialData = ', 0, '};') + '}')
+    section = next(search_dict(data, 'itemSectionRenderer'), None)
+    renderer = next(search_dict(section, 'continuationItemRenderer'), None) if section else None
+    if not renderer:
+        return
+
+    needs_sorting = sort_by != SORT_BY_POPULAR
```

Comment:
What's the purpose of `needs_sorting`?

Reply:
Sorted the comments if it's not sorted by popularity. (By latest comment)